### PR TITLE
Add canning to the ChemMaster

### PR DESCRIPTION
**Pull request by TDHooligan**
## About the PR 
Allows players to create cans in the ChemMaster and choose their design from the existing set

## Why's this needed? 
More fun gimmicks! Custom soda brands, etc.

## Changelog

```changelog
(u)TDHooligan
(+)Soda cans can now be printed in the CheMaster 3000, start a brand!
```

### CODE REVIEW

**Comment by DimWhat** [comment]
Kinda a balance thing considering soda cans are 50u and vials (which the chemicompiler can currently produce) contain 5u of chemicals.

**Comment by TDHooligan** [comment]
> Kinda a balance thing considering soda cans are 50u and vials (which the chemicompiler can currently produce) contain 5u of chemicals.

this is for the chemmaster

**Comment by DimWhat** [comment]
I see, well ignore what I said I thought this was still about chemicompiler

File: code/modules/chemistry/Chemistry-Machinery.dm [hunk 1]
```diff
@@ -483,6 +483,26 @@
...
+			var/default = R.get_master_reagent_name()
+			var/input_name = input(usr, "Name the can:", "Name", default) as null|text
+			if(input_name && input_name != default)
+				phrase_log.log_phrase("bottle", input_name, no_duplicates=TRUE)
+			var/bottlename = copytext(html_encode(input_name), 1, 32)
+
+			var/input_design = input(usr, "Choose the design (1~26):", "Design", default) as null|num
+
+			if (isnull(bottlename) || !src.beaker || !R || !length(bottlename) || bottlename == " " || get_dist(usr, src) > 1 || isnull(input_design) || input_design > 26 || input_design < 1)
+				return
```
Comment (Azrun) [inline code/modules/chemistry/Chemistry-Machinery.dm:496]:
```suggestion
			if (!src.beaker || !R || !length(bottlename) || bottlename == " " || get_dist(usr, src) > 1 || isnull(input_design) || input_design > 26 || input_design < 1)
				return
```
`length(null)` is 0

**Comment by Azrun** [comment]
I'm not sure if maybe it should be limited to the more generic cans or some other way of being identifiable as custom.  The inherent risk of eating pills off the floor is more intuitive than finding a sealed can disguised as a standard item. 🤔 

I do like the _flavor_ this could add to gimmicks though.

**Comment by Kyle2143** [comment]
I might be misunderstanding something here, it's hars to tell without pictures and just by looking at the changes, but do these bottles/cans that it creates have any visual indication of what the reagent inside is? Like how pills and bottle liquid are the reagents' colors mixed together?

I realize it wouldn't make sense to see the reagent through an opaque metal can, but I think it should still have a visual indicator, like have

**Comment by TDHooligan** [comment]
the cans have custom identify text for those who how to check, plus the existing soda cans are also rarely seen around. My personal experience was that I didn't even know the random ingredients in an 'off-brand' soda were usually harmless food chems and refused to trust them.

I dunno, I feel newbies drinking a sip and bursting into flames once or twice will quickly cause the 'floorcan' effect, but you might see some more creative med/science endeavors with them looking more official than just pills.

File: code/modules/chemistry/Chemistry-Machinery.dm (hunk 1 above)
Comment (TDHooligan) [reply code/modules/chemistry/Chemistry-Machinery.dm:496]: happy to take it out if you want, it's identical with the bottling code i lazily copied over and it seems a little dodgy to rely on the length of nulls, that's just the C# talkin though

**Review by Azrun — APPROVED** [review]
Looks good.  

Could be neat if the designs were limited based on the source volume similar to `"createbottle"`, such that the volume of the source could allow for small/normal/large cans.  But that kind of sounds like feature creep at the moment.

**Comment by TDHooligan** [comment]
i've made sprites for small/normal variants of each, i think adding 'large' for a lot of the designs is awkward for some and pointless for others. will upload later today if the PR isn't merged

**Comment by TDHooligan** [comment]
image

**Review by Azrun — APPROVED** [review]

**Comment by Kyle2143** [comment]
Oh, those're pretty fun.

**Pull request merged**